src/tool_operate.c

```diff
@@ -656,16 +656,106 @@ static void single_transfer_cleanup(struct OperationConfig *config)
   }
 }
 
+static char *ipfs_gateway(void)
+{
+  char *gateway;
+  char *ipfs_path;
+  char *gateway_file_path;
+  FILE *gateway_file;
+
+  gateway = curlx_getenv("IPFS_GATEWAY");
+
+  /* Gateway is found from environment variable. */
+  if(gateway) {
+    bool add_slash = (gateway[strlen(gateway) - 1] == '/') ? FALSE : TRUE;
+    gateway = aprintf("%s%s", gateway, (add_slash) ? "/" : "");
+    return gateway;
+  }
+
+  /* Try to find the gateway in the IPFS data folder. */
+
+  ipfs_path = curlx_getenv("IPFS_PATH");
+
+  if(!ipfs_path) {
+    /* Empty path, fallback to "~/.ipfs", as that's the default location. */
+    ipfs_path = aprintf("%s/.ipfs/", curlx_getenv("HOME"));
+  }
+
+  gateway_file_path = aprintf("%sgateway", ipfs_path);
+  gateway_file = fopen(gateway_file_path, FOPEN_READTEXT);
+
+  if(gateway_file) {
+    char gateway_buffer[255];
+    if(!fgets(gateway_buffer, sizeof(gateway_buffer) - 1, gateway_file)) {
+      fclose(gateway_file);
+      return NULL;
+    }
+    fclose(gateway_file);
+
+    /* Replace the last char with \0 */
+    gateway_buffer[sizeof(gateway_buffer) - 1] = 0;
+
+    /* Replace first occurence of end of line with \0 */
+    gateway_buffer[strcspn(gateway_buffer, "\r")] = 0;
+    gateway_buffer[strcspn(gateway_buffer, "\n")] = 0;
+
+    if(strlen(gateway_buffer) > 1) {
+      bool add_slash = (gateway_buffer[strlen(gateway_buffer) - 1] == '/')
+        ? FALSE
+        : TRUE;
+
+      gateway = aprintf("%s%s", gateway_buffer, (add_slash) ? "/" : "");
+      return gateway;
+    }
+  }
+
+  return NULL;
+}
+
+/*
+ * Rewrite ipfs://<cid> and ipns://<cid> to be a normal
+ * URL that can be handled by an IPFS gateway.
+ */
+static void ipfs_url_rewrite(char **url)
+{
+  char *inner = *url;
+  char *gateway;
+  const char *protocol;
+  char *everything_after_scheme;
+  char *urlbuffer;
+
+  /* 7 is "ipfs://" and "ipns://", it needs o be at least 7 characters*/
+  if(strlen(inner) < 7) {
```

Comment:
You already checked the scheme when this function is invoked, do you really need to add an extra (inferior because you shouldn't assume URL syntax) check here?

Reply:
Agreed. I do need some checks but this one is perhaps a bit too "loose".